include/revng/Pipeline/Pipe.h

```diff
@@ -0,0 +1,311 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+#include <set>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Error.h"
+
+#include "revng/ADT/STLExtras.h"
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/InputOutputContract.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/FunctionIntrospection.h"
+
+/// A pipe is a class with the following characteristics
+/// * It must have a static constexpr field named Name that is a string
+///   describing its name. Mostly used for debug purposes.
+///
+/// * A std::array<AtomicContract, X> getContract() const method that returns
+///   the contract of such class.
+///
+/// * a void run(T...) method where the first argument type can optionally be
+///   Context and every other type must be the most derived type of a Container.
+///   The operation performed by run MUST be consisted with the contract.
+///
+///
+
+namespace Pipeline {
+
+/// A simple pair name pointer used to attach the name of a container
+/// declaration to the container using it.
+///
+/// This is usefull when a pipe needs to know the name of the container.
+template<typename T>
+class NamedPointer {
```

Comment:
The name should reflect the fact that this is designed to be used with containers, as the doc says.